core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -1986,6 +1986,112 @@ private[spark] object Utils extends Logging {
   }
 
   /**
+   * Attempt to start a service using a given list of port ranges, or fail after a number of attempts.
+   * Each subsequent attempt uses the next available port within the specified set of ranges.
+   * @param startPort the port to check first
+   * @param startService Function to start service on a given port.
+   *                     This is expected to throw java.net.BindException on port collision.
+   * @param conf A SparkConf used to get the maximum number of retries when binding to a port.
+   * @param serviceName Name of the service.
+   * @return (service: T, port: Int)
+   */
+  def startServiceFromPortRange[T](
```

Comment:
At first glance there's too much duplicated code for this to go in. I think it would be good to find a way to refactor existing code without duplicating so much of it.